mesonbuild/environment.py

```diff
@@ -1460,8 +1481,8 @@ def can_run(self):
 class PerMachineDefaultable(PerMachine):
     """Extends `PerMachine` with the ability to default from `None`s.
     """
-    def __init__(self):
-        super().__init__(None, None, None)
+    def __init__(self, build=None, host=None, target=None):
```

Comment:
This isn't a very good idea, because `None` is the sentinal value used to indicate an uninitialized one. Your use above would be incorrect then, because the host one is never null so `default_missing` is a no-opt.

In other words, we have spomething like (I'll just use Rust-like syntax for now):
```python
class PerMachine[T]:
  ...
class PerMachineDefaultable[T](PerMachine[T]):
  def __init__(self, build: NullOr[T] =None, host: NullOr[T]=None, target: NullOr[T]=None):
```

Reply:
I honestly don't like the design of the `PerMachineDefaultable` class. Setting member is one of the few things that the constructor of a well behaved class should do (not that meson has well behaved classes :disappointed:). I dislike that the correct usage is:
```python
a = PerMachineDefaultable()
if True:
    a.host = Foo()
if False:
   a.build = Bar()
a.default_missing()
```
It's a little hard for me to exactly what I dislike about it, but there are two things: We expose the sentinel value (which presents the chance of a TypeError), and that you have to set required instance members outside of the class. 

What if we redesigned the class like this (I've left out inheritance to simplify the example):
```python
from typing_extensions import Protocol
import typing

class BooleanProtocol(Protocol):
    def __bool__(self) -> bool: ...

B = typing.TypeVar('B', bound=BooleanProtocol)

class PerMachineDefault(typing.Generic[B]):
    def __init__(self, factory: typing.Callable[[], B]) -> None:
        self.build: B = factory()
        self.host: B = factory()
        self.target: B = factory()

    def default_missing(self):
        if not self.host:
             self.host = self.build
        if not self.target:
            self.target = self.host
```

The advantage of this design is that a) we never expose a sentinel value so there's never a chance of a `TypeError`, and 2 the internal types (though mutable) are set in the constructor. For well behaved containers this automatically does the right thing, set the factory to `dict`, `list`, `set`, etc and everything just works.

Reply:
You really do want the None because cardinality wise there should be an extra value to indicate uninitialized which is impossible after.

I think better would be to make `PerMachineDefaultable` `PerMachineFactory`. Then `default_missing` becomes `build`, giving you a `PerMachine`. `PerMachine` itself has no nullable fields.

Reply:
making the factory a local variable, and only doing `self.paths = perMachineFactor.build()` also makes it "more transactional" in that the long-lived thing isn't mutate but only initialized once.

Reply:
But before you shave that yak, look at `environment.machines` and try not to puke. In the least declarative case, the inferred machine is used to find a compiler which tells us more about that same machine (or even contradictory information!), so there's more mutation no matter what.